### ISSUE TITLE
Async?

### ISSUE THREAD

**Issue by bain3** [bain3/pronotepy#262]
I am thinking about `3.0`, and if the library should be async, or not. What are your thoughts? React with :+1:, or :-1:, please.

**Comment by bain3** [comment]
What I have so far:

- [ ] clean up parsing, split `dataClasses` file, since it is becoming a mess
- [ ] try to use the [`dataclasses`](https://docs.python.org/3/library/dataclasses.html) module.
- [ ] make all the properties (like [`Period.grades`](https://pronotepy.readthedocs.io/en/stable/api/data.html#pronotepy.Period.grades)) functions, since they actually make requests, which is kinda hidden
- [ ] don't [calculate averages](https://github.com/bain3/pronotepy/blob/ccfcfd539d7d6c2ccb9b957d02fe57d27248a18d/pronotepy/dataClasses.py#L547) if there is none, just stick to what pronote gives us
- [ ] keep communication logic, login, etc... inside the communication class. The client should really just make requests, not caring about encryption, or how to log in.

**Comment by tarneaux** [comment]
I'd also add a general code cleanup.
For example, why do both `teacher_name` and `teacher_names` exist for the same [`pronotepy.lesson`](https://pronotepy.readthedocs.io/en/stable/api/data.html#pronotepy.Lesson) dataclass? Shouldn't one of them be removed? Is `Lesson.virtual_classrooms` still relevant (does pronote still have that feature)?

**Comment by bain3** [comment]
Yes, definitely. The `teacher_name` attribute is for [backwards compatibility](https://github.com/bain3/pronotepy/blob/ccfcfd539d7d6c2ccb9b957d02fe57d27248a18d/pronotepy/dataClasses.py#L939). There are multiple things that should be removed, but were left there, so that we don't break the API unnecessarily.

**Comment by bain3** [comment]
I have started working on `v3`. You can find the [documentation here](https://bain.cz/t/v3docs/). Feel free to message me if you encounter any issues, or have any other thoughts. Anything can still change.
